error-prone: implement BanClassLoader with forbidden-apis instead

rmuir opened the pull request:
This rule is attempting to ban dangerous usages of ClassLoader which could be security hazard (IMO a good idea). forbidden APIs is a good tool for banning, just like we use forbidden-apis to ban JNDI and other things.

rmuir committed:
error-prone: implement BanClassLoader with forbidden-apis instead

This rule is attempting to ban dangerous usages of ClassLoader which
could be security hazard (IMO a good idea). forbidden APIs is a good
tool for banning, just like we use forbidden-apis to ban JNDI and other
things.

rmuir commented on gradle/validation/forbidden-apis/defaults.all.txt:
@uschindler if you have any ideas, i'd love to fix this. This is just using the same banned list that error-prone uses, except as forbidden signatures.

uschindler replied:
In reality most of the signatures here are automatically enabled by default (in a more dynamic way).

- `jdk.internal.*` is forbidden by default by disallowing all calls to non "public access" modules. See the "jdk-non-portable" bundled signatures: https://github.com/policeman-tools/forbidden-apis/wiki/BundledSignatures
- `sun.misc.Unsafe` is also forbidden by default (same as before: https://github.com/policeman-tools/forbidden-apis/wiki/BundledSignatures). Adding it here brings more problems than it solves. Because `sun.misc.Unsafe` get removed around Java 25 (and especailly that method mentioned here), this will cause issues in later Java versions.

Because of that all "internal" and "sun.misc" stuff should be removed here. The other ones are fine,

uschindler replied:
All fine. I think thats documented somewhere. But basically the signatures files need the "binary name" of all classes.
I will add the [wiki](https://github.com/policeman-tools/forbidden-apis/wiki/SignaturesSyntax) page to explicitely give an example (in addition to `java.lang.String`).

rmuir committed:
tidy

uschindler commented:
> Source of the check: https://github.com/google/error-prone/blob/master/core/src/main/java/com/google/errorprone/bugpatterns/BanClassLoader.java
> 
> It also checks for explicit `extends URLClassLoader`, which isn't handled here. I'm not sure what special powers subclassing gives, maybe more methods need to be banned.
> 
> Also maybe some of these don't need to be banned because they are impossible due to java module system? I think google is still on java 8 :)

Exactly, see above. Most of the signatures here can be removed. jdk.internal is unreachable unless you pass open-modules parameters to jvaca or jdk. Nevertheless to SUpport Java 8, the forbiddenapis has a dynamic `jdk-non-portable` bundled signature which disallows all non-documented classes like sun.misc or those in `jdk.*` java modules.

About subclassing: A constructor is like a static methos. To forbid a constructur you need to give exact class names and ctors, there's no inheritance.

uschindler commented:
Final note abozut subclasses: You need to add all ctors of sublasses of Classloader which reside in JDK. Now you might argue that one could define a subclass in user code (extending URLClassLoader) to circumvent the check. But this is differeent because when you declare that "user defined" classloader you have to add a ctor which calls super -> then it detect calling the call to superclass' ctor which was forbidden.

rmuir committed:
simplify rules and fix MethodHandles$Lookup

Remove internals which are already banned
Fix syntax for MethodHandles$Lookup
Simplify RMI server (we can ban the class outright)

rmuir committed:
improve check as goal is to detect turning byte[] into code
